### PR TITLE
postinst: hardcode python binary and PYTHONPATH

### PR DESCRIPTION
**Pull request by orndorffgrant**
## Proposed Commit Message

```
postinst: hardcode python binary and PYTHONPATH

When the user's environment is set up to use a nonstandard python
installation, our postinst would fail. By specifying the exact python
binary to use, we avoid this problem. By also setting PYTHONPATH to the
location where our uaclient library is installed, we increase the
chances that we will succeed even if /usr/bin/python3 is overwritten.

If /usr/bin/python3 is overwritten with a different version of python,
whose standard library is incompatible with the expected version of
python, it is possible postinst could still fail.

Separately, if /usr/bin/python3 is overwritten with a different version
of python, the ua executable _will_ still fail.

LP: #1946034 #1944700 #1930121 #1938290 #1938074 #1927905
```

## Test Steps

1. Run the new integration tests.
2. Forgive me for making Jenkins compile python from source for CI (This is the only "natural" way that I could find where user will end up with a nonstandard python install that causes the symptoms of the linked LP bugs).

## Desired commit type

 - [ ] Squash and merge: Using the "Proposed Commit Message"
 - [ ] Create a merge commit and use "Proposed Commit Message"
 - [x] Rebase and merge, no merge commit, rely only on existing commit messages

## Checklist

 - [ ] I have updated or added any unit tests accordingly
 - [x] I have updated or added any integration tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

File: debian/ubuntu-advantage-tools.postinst [hunk 1]
```diff
@@ -16,6 +16,9 @@ fi
 # https://lintian.debian.org/tags/postinst-does-not-load-confmodule.html
 . /usr/share/debconf/confmodule
 
+PYTHON=/usr/bin/python3
+export PYTHONPATH=/usr/lib/python3/dist-packages
```
[the remark this answers is not in the document]
Comment (orndorffgrant) [reply debian/ubuntu-advantage-tools.postinst:20]: This should only affect commands run as part of this script, but that is a good point that that is unnecessary, we could instead change this to
```
PYTHON=/usr/bin/env PYTHONPATH=/usr/lib/python3/dist-packages /usr/bin/python3
```

and then we wouldn't need the export line. Thoughts?

File: debian/ubuntu-advantage-tools.postinst [hunk 2]
```diff
@@ -16,6 +16,8 @@ fi
 # https://lintian.debian.org/tags/postinst-does-not-load-confmodule.html
 . /usr/share/debconf/confmodule
 
+PYTHON="/usr/bin/env PYTHONPATH=/usr/lib/python3/dist-packages /usr/bin/python3"
```
Comment (lucasmoura) [inline debian/ubuntu-advantage-tools.postinst:19]: Can we only have `/usr/lib/python3/`  as the python path here ?

File: features/util.py [hunk 3]
```diff
@@ -113,13 +114,24 @@ def repo_state_hash(include_features: bool = False):
...
+
     git_diff_cmd = ["git", "diff"]
     if not include_features:
         files_and_folders = [f for f in os.listdir(".") if "features" not in f]
         git_diff_cmd += ["--", *files_and_folders]
-    rev_out = subprocess.check_output(git_rev_cmd)
     diff_out = subprocess.check_output(git_diff_cmd)
-    output_to_hash = rev_out + diff_out
+
+    git_status_cmd = ["git", "status", "--porcelain"]
```
Comment (lucasmoura) [inline features/util.py:125]: I see why we are doing this, but I don't know if we should. Any test script that we create or any manual test we add, will trigger for a new package build. I know this is the more complete solution, but I don't know if it pays off.

What do you think @renanrodrigo @blackboxsw ?

Comment (orndorffgrant) [reply features/util.py:129]: I updated this function to exclude more directories

File: debian/ubuntu-advantage-tools.postinst (hunk 2 above)
Comment (orndorffgrant) [reply debian/ubuntu-advantage-tools.postinst:19]: I just tested and unfortunately, having only `/usr/lib/python3/` doesn't work here

File: features/util.py [hunk 4]
```diff
@@ -88,38 +88,61 @@ def lxc_get_property(name: str, property_name: str, image: bool = False):
         return value
 
 
-def repo_state_hash(include_features: bool = False):
+def repo_state_hash(
+    exclude_dirs: List[str] = [".github", "demo", "features", "sru", "tools"]
```
Comment (lucasmoura) [inline features/util.py:92]: Even though we are not doing this at the moment, but if we modify the `exclude_dirs` default list on the function, we will modify it on all other calls of `repo_state_hash`.

I know we are not doing that, but it is better to be defensive here. It is better to have that as `None`  by default and set it up with a default list.

File: features/util.py [hunk 5]
```diff
@@ -88,38 +88,61 @@ def lxc_get_property(name: str, property_name: str, image: bool = False):
...
+    git_status_cmd = ["git", "status", "--porcelain"]
+    status_out = subprocess.check_output(git_status_cmd).decode("utf-8")
+    status_lines = status_out.split("\n")
+    new_files = [name[3:] for name in status_lines if name.startswith("?? ")]
+    new_file_content = ""
+    for fname in new_files:
+        exclude = False
+        for exclude_dir in exclude_dirs:
+            if fname.startswith(exclude_dir):
+                exclude = True
```
Comment (lucasmoura) [inline features/util.py:139]: Let's add a `break` here

File: features/util.py (hunk 3 above)
Comment (renanrodrigo) [reply features/util.py:130]: I think it is fine. We can tweak the "watched" directories in the future if needed, but so far, it works

**Review by lucasmoura — APPROVED** [review]

**Comment by orndorffgrant** [comment]
LP 1920121 is now the single bug for this issue.

I've updated the commit and fixed the mypy issue that stopped the last jenkins run. Let's let jenkins do a full run now.

**Pull request merged**